sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala

```diff
@@ -470,3 +475,283 @@ class StreamingJoinSuite extends StreamTest with StateStoreMetricsTest with Befo
+    val stream = MemoryStream[Int]
+      .toDF()
+      .withColumn("timestamp", 'value.cast("timestamp"))
+      .withWatermark("timestamp", "1 second")
+    val joined =
+      stream.join(Seq(1).toDF(), Seq("value"), "left_outer")
+
+    // This test is in the suite just to confirm the validations below don't block this valid join.
+    // We don't need to check results, just that the join can happen.
+    testStream(joined)()
+  }
+
+  test("left batch stream outer join") {
+    val stream = MemoryStream[Int]
+      .toDF()
+      .withColumn("timestamp", 'value.cast("timestamp"))
+      .withWatermark("timestamp", "1 second")
+    val joined =
+      Seq(1).toDF().join(stream, Seq("value"), "left_outer")
+
+    val thrown = intercept[AnalysisException] {
+      testStream(joined)()
+    }
+
+    assert(thrown.getMessage.contains(
+      "Left outer join with a streaming DataFrame/Dataset on the right and a static"))
+  }
+
+  test("right stream batch outer join") {
+    val stream = MemoryStream[Int]
+      .toDF()
+      .withColumn("timestamp", 'value.cast("timestamp"))
+      .withWatermark("timestamp", "1 second")
+    val joined =
+      stream.join(Seq(1).toDF(), Seq("value"), "right_outer")
+
+    val thrown = intercept[AnalysisException] {
+      testStream(joined)()
+    }
+
+    assert(thrown.getMessage.contains(
+      "Right outer join with a streaming DataFrame/Dataset on the left and a static"))
+  }
+
+  test("left outer join with no watermark") {
+    val joined =
+      MemoryStream[Int].toDF().join(MemoryStream[Int].toDF(), Seq("value"), "left_outer")
+
+    val thrown = intercept[AnalysisException] {
+      testStream(joined)()
+    }
+
+    assert(thrown.getMessage.contains(
+      "Stream-stream outer join between two streaming DataFrame/Datasets is not supported " +
+        "without a watermark"))
+  }
+
+  test("right outer join with no watermark") {
+    val joined =
+      MemoryStream[Int].toDF().join(MemoryStream[Int].toDF(), Seq("value"), "right_outer")
+
+    val thrown = intercept[AnalysisException] {
+      testStream(joined)()
+    }
+
+    assert(thrown.getMessage.contains(
+      "Stream-stream outer join between two streaming DataFrame/Datasets is not supported " +
+        "without a watermark"))
+  }
+
+  test("windowed left outer join") {
+    val (leftInput, rightInput, joined) = setupWindowedJoin("left_outer")
+
+    testStream(joined)(
+      // Test inner part of the join.
+      AddData(leftInput, 1, 2, 3, 4, 5),
+      AddData(rightInput, 3, 4, 5, 6, 7),
+      CheckLastBatch((3, 10, 6, 9), (4, 10, 8, 12), (5, 10, 10, 15)),
+      // Old state doesn't get dropped until the batch *after* it gets introduced, so the
+      // nulls won't show up until the next batch after the watermark advances.
+      AddData(leftInput, 21),
+      AddData(rightInput, 22),
+      CheckLastBatch(),
+      AddData(leftInput, 22),
+      CheckLastBatch(Row(22, 30, 44, 66), Row(1, 10, 2, null), Row(2, 10, 4, null))
```

Comment:
assert on the number of state rows, similar to the inner join tests. 
for the below tests as well. this is to ensure that the state is actually being dropped.